[shard] Extensible Sharder and ShardingPlanner

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #75844



Add extensible Sharder and ShardingPlanner for advanced users to extend and do custom sharding, some details:

* ShardingPlanner is merely a interface that could be extended to plan advanced sharding strategies, it will produce a ShardingPlan that outlines the detailed sharding strategies for each submodule, note that it could also include custom sharders.
* a custom Sharder is another interface that could be extended to define advanced sharding strategies for specific modules (i.e. a collection of `nn.EmbeddingBag`s.) The custom sharder defined by the user can be picked up by `share_module` API. `shard_module` will call `Sharder.shard` to the module, the output of the `Sharder.shard` will replace the original submodule.


Differential Revision: [D35658657](https://our.internmc.facebook.com/intern/diff/D35658657/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35658657/)!

Differential Revision: [D35658657](https://our.internmc.facebook.com/intern/diff/D35658657)

Commit:
[WIP][shard] Extensible Sharder and ShardingPlanner

Add extensible Sharder and ShardingPlanner for advanced users to extend and do custom sharding

Differential Revision: [D35658657](https://our.internmc.facebook.com/intern/diff/D35658657/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35658657/)!

[ghstack-poisoned]

Commit:
Update on "[shard] Extensible Sharder and ShardingPlanner"




Add extensible Sharder and ShardingPlanner for advanced users to extend and do custom sharding, some details:

* ShardingPlanner is merely a interface that could be extended to plan advanced sharding strategies, it will produce a ShardingPlan that outlines the detailed sharding strategies for each submodule, note that it could also include custom sharders.
* a custom Sharder is another interface that could be extended to define advanced sharding strategies for specific modules (i.e. a collection of `nn.EmbeddingBag`s.) The custom sharder defined by the user can be picked up by `share_module` API. `shard_module` will call `Sharder.shard` to the module, the output of the `Sharder.shard` will replace the original submodule.


Differential Revision: [D35658657](https://our.internmc.facebook.com/intern/diff/D35658657/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35658657/)!

Differential Revision: [D35658657](https://our.internmc.facebook.com/intern/diff/D35658657)

[ghstack-poisoned]

Commit:
Update on "[shard] Extensible Sharder and ShardingPlanner"




Add extensible Sharder and ShardingPlanner for advanced users to extend and do custom sharding, some details:

* ShardingPlanner is merely a interface that could be extended to plan advanced sharding strategies, it will produce a ShardingPlan that outlines the detailed sharding strategies for each submodule, note that it could also include custom sharders.
* a custom Sharder is another interface that could be extended to define advanced sharding strategies for specific modules (i.e. a collection of `nn.EmbeddingBag`s.) The custom sharder defined by the user can be picked up by `share_module` API. `shard_module` will call `Sharder.shard` to the module, the output of the `Sharder.shard` will replace the original submodule.


Differential Revision: [D35658657](https://our.internmc.facebook.com/intern/diff/D35658657/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35658657/)!

Differential Revision: [D35658657](https://our.internmc.facebook.com/intern/diff/D35658657)

[ghstack-poisoned]

pritamdamania87 commented on test/distributed/_shard/test_sharder.py:
```diff
@@ -0,0 +1,151 @@
+    @with_comms(init_rpc=False)
+    @skip_if_lt_x_gpu(TEST_GPU_NUM)
+    @requires_nccl()
+    def test_custom_sharder_errors(self):
```
Can we also test the case where a custom sharder and an entry in the sharding plan conflict? We should error out in such cases.

pritamdamania87 commented on torch/distributed/_shard/api.py:
```diff
@@ -265,12 +266,21 @@ def shard_module(
                 src_rank=src_rank,
                 process_group=process_group
             )
+        elif isinstance(spec, Sharder):
```
We need some validation here that another item in ShardingPlan doesn't conflict with the submodule tree that Sharder is working with.

pritamdamania87 commented on torch/distributed/_shard/sharding_plan/api.py:
```diff
@@ -51,6 +55,26 @@ class ShardingPlan(object):
         >>> )
     """
-    plan: Dict[str, ShardingSpec]
+    plan: Dict[str, Union[ShardingSpec, Sharder]]
```
Need to update the docs to reflect this too

pritamdamania87 commented on torch/distributed/_shard/sharder.py:
```diff
@@ -0,0 +1,27 @@
+                The module to apply sharding to.
+        Returns:
+            A :class:`torch.nn.Module` object that represents a module
+            that's already been sharded (i.e. contains ShardedTensor parameters).
```
> (i.e. contains ShardedTensor parameters).

This might not always be the case since the user might do their own thing and not really use ShardedTensor.

Commit:
Update on "[shard] Extensible Sharder and ShardingPlanner"




Add extensible Sharder and ShardingPlanner for advanced users to extend and do custom sharding, some details:

* ShardingPlanner is merely a interface that could be extended to plan advanced sharding strategies, it will produce a ShardingPlan that outlines the detailed sharding strategies for each submodule, note that it could also include custom sharders.
* a custom Sharder is another interface that could be extended to define advanced sharding strategies for specific modules (i.e. a collection of `nn.EmbeddingBag`s.) The custom sharder defined by the user can be picked up by `share_module` API. `shard_module` will call `Sharder.shard` to the module, the output of the `Sharder.shard` will replace the original submodule.


Differential Revision: [D35658657](https://our.internmc.facebook.com/intern/diff/D35658657/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35658657/)!

Differential Revision: [D35658657](https://our.internmc.facebook.com/intern/diff/D35658657)

[ghstack-poisoned]

Commit:
Update on "[shard] Extensible Sharder and ShardingPlanner"




Add extensible Sharder and ShardingPlanner for advanced users to extend and do custom sharding, some details:

* ShardingPlanner is merely a interface that could be extended to plan advanced sharding strategies, it will produce a ShardingPlan that outlines the detailed sharding strategies for each submodule, note that it could also include custom sharders.
* a custom Sharder is another interface that could be extended to define advanced sharding strategies for specific modules (i.e. a collection of `nn.EmbeddingBag`s.) The custom sharder defined by the user can be picked up by `share_module` API. `shard_module` will call `Sharder.shard` to the module, the output of the `Sharder.shard` will replace the original submodule.


Differential Revision: [D35658657](https://our.internmc.facebook.com/intern/diff/D35658657/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35658657/)!

Differential Revision: [D35658657](https://our.internmc.facebook.com/intern/diff/D35658657)

[ghstack-poisoned]
